larpandoracontent/LArThreeDReco/LArCosmicRay/TwoViewDeltaRayMatchingAlgorithm.cc

```diff
@@ -0,0 +1,466 @@
+
+    auto iter(hitTypeToIndexMap.find(hitType));
+
+    if (iter == hitTypeToIndexMap.end())
+      return element.GetOverlapResult().GetBestMatchedAvailableCluster(); 
+
+    if (iter->second == 1)
+        return element.GetCluster1();
+
+    return element.GetCluster2();
+}
+    
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+bool TwoViewDeltaRayMatchingAlgorithm::DoesClusterPassTensorThreshold(const Cluster *const pCluster) const
+{
+    return (pCluster->GetNCaloHits() >= m_minClusterCaloHits);
+}    
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+    
+void TwoViewDeltaRayMatchingAlgorithm::CalculateOverlapResult(const Cluster *const pCluster1, const Cluster *const pCluster2, const Cluster *const)
+{
+    TwoViewDeltaRayOverlapResult overlapResult;
+    PANDORA_THROW_RESULT_IF_AND_IF(STATUS_CODE_SUCCESS, STATUS_CODE_NOT_FOUND, !=, this->CalculateOverlapResult(pCluster1, pCluster2, overlapResult));
+    
+    if (overlapResult.IsInitialized())
+        this->GetMatchingControl().GetOverlapMatrix().SetOverlapResult(pCluster1, pCluster2, overlapResult);
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------     
+
+StatusCode TwoViewDeltaRayMatchingAlgorithm::CalculateOverlapResult(const Cluster *const pCluster1, const Cluster *const pCluster2, TwoViewDeltaRayOverlapResult &overlapResult) const
+{
+    float xMin1(0.f), xMax1(0.f), xMin2(0.f), xMax2(0.f);
+    pCluster1->GetClusterSpanX(xMin1, xMax1); 
+    pCluster2->GetClusterSpanX(xMin2, xMax2); 
+
+    const float xOverlap(std::min(xMax1, xMax2) - std::max(xMin1, xMin2));
+
+    if (xOverlap < std::numeric_limits<float>::epsilon())
+        return STATUS_CODE_NOT_FOUND;
+
+    PfoList commonMuonPfoList;
+    this->FindCommonMuonParents(pCluster1, pCluster2, commonMuonPfoList);
+    
+    if (commonMuonPfoList.empty())
+        return STATUS_CODE_NOT_FOUND;
+
+    CartesianPointVector projectedPositions;
+    StatusCode status(this->GetProjectedPositions(pCluster1, pCluster2, projectedPositions));
+    
+    if (status != STATUS_CODE_SUCCESS)
+        return status;
+
+    // Find all matched clusters (including unavailable)
+    ClusterList matchedClusterList;
+    this->CollectThirdViewClusters(pCluster1, pCluster2, projectedPositions, matchedClusterList);
+
+    if (matchedClusterList.empty())
+        return STATUS_CODE_NOT_FOUND;
+    
+    const Cluster *pBestMatchedCluster(nullptr); 
+    float reducedChiSquared(std::numeric_limits<float>::max());
+
+    this->GetBestMatchedCluster(pCluster1, pCluster2, commonMuonPfoList, matchedClusterList, pBestMatchedCluster, reducedChiSquared);
+
+    //ATTN: Ignore if other clusters matches have more hits
```

Comment:
great question! so, the TwoViewDeltaRayMatching alg is essentially trying to pick up fragments in the third view (or allow clusters in two views to be matched if one understands that in the third view the cosmic ray is hiding the delta ray or one sees that the third view of the match has already been taken) but it's the first thing that is important here. So the alg runs three times, each time looking in a different view. Essentially when I create the match I want to do so with the best pair of clusters since this will give me the biggest projection. So if i find that my best matched projected cluster has more hits then one of the clusters i use to do the projection that is the one i want to form my pair. So if i find this, I leave the pair and don't put it in the matrix with the idea that when the next version of the alg runs it will pick up the pair and do the projecting with the best pair. It's important to make sure the 'pBestMatchedCluster' is available, you see it could be a muon so then this would always win the 'highest hit' contest which I don't really want to happen